test/unit/cmap/connection.test.ts

```diff
@@ -142,4 +146,181 @@ describe('new Connection()', function () {
     expect(beforeHandshakeSymbol).to.be.a('symbol');
     expect(error).to.have.property(beforeHandshakeSymbol, true);
   });
+
+  describe('NODE-6370: regression test', function () {
+    class MockSocket extends Socket {
+      override write(_data: string | Buffer) {
+        return false;
+      }
+    }
+
+    let socket: MockSocket;
+    let connection: Connection;
+
+    this.timeout(10_000);
+
+    beforeEach(function () {
+      socket = new MockSocket();
+      connection = new Connection(socket, {});
+    });
+
+    const validResponse = Buffer.from(
+      'a30000002a0800004b010000dd07000000000000008e000000016f6b00000000000000f03f0324636c757374657254696d65005800000011636c757374657254696d65001c00000093f6f266037369676e61747572650033000000056861736800140000000072d8d6eab4e0703d2d50846e2db7adb5d2733cc4126b65794964000200000026f6f2660000116f7065726174696f6e54696d65001c00000093f6f26600',
+      'hex'
+    );
+
+    const chunks = [validResponse.slice(0, 10), validResponse.slice(10)];
+
+    describe('when data is emitted before drain', function () {
```

Comment:
If the changes in `connection.ts` are reverted, most of these tests pass.  

1. all the `first command` tests pass because the message stream starts paused.
2. the `when there is a delay between data and drain` test in the not first command section fails.  
3. everything else passes.

I did some debugging, and determined that there is some async delay after the message stream calls its callback with the full message and when the message stream emits a data event.  without digging further into Nodejs source, I'm not sure 1. if this is expected and 2. why this is.

I'm not convinced it matters for these tests - I think that there's adequate coverage of the various scenarios here that is resilient to changes in async behavior.

Reply:
If I understand correctly, that is a guarantee. The stream makes sure the callback is run "nextTick" (might use a different API) so that you don't get an inconsistent execution experience. Otherwise sometimes (ex. `write(.., cb); nextLine;`) `cb` would run before `nextLine`

I agree, this coverage is worth having, maybe future stream implementations or different runtimes could introduce this scenario and we're defended against it

Reply:
The `callback` that users provide to `write()` is not the same as the callback applied in `transform` (see [this](https://nodejs.org/api/stream.html#transform_transformchunk-encoding-callback:~:text=The%20callback%20function%20must%20be%20called%20only,In%20other%20words%2C%20the%20following%20are%20equivalent%3A)).  

Also, I think that for transform streams the behavior you've mentioned is not guaranteed - it would depend on whether the implementation of `write` is synchronous or not (there's no guarantee, best I can tell: https://nodejs.org/api/stream.html#transform_transformchunk-encoding-callback:~:text=transform._transform()%20is%20never%20called%20in%20parallel%3B%20streams%20implement%20a%20queue%20mechanism%2C%20and%20to%20receive%20the%20next%20chunk%2C%20callback%20must%20be%20called%2C%20either%20synchronously%20or%20asynchronously.). 

What you can rely on being true is that the ordering of the following is strictly defined:

```typescript
transform.write(chunk, () => { 
	nextLine;
});
```